docs/DeveloperGuide.md

```diff
@@ -68,10 +68,84 @@ The `Model`,
+The Logic initialises the `BinaryTree<Gate>` object using the parameterised constructor, thus specifying the Gate class type 
+root. The object diagram below depicts the initial state of the Model when a `BinaryTree<Gate>` object is created.
+
+![InitialBinaryTree](diagrams/BinaryTreeInitialObjectDiagram.png)
+
+The Logic uses the parameterised constructor of `BinaryTree<T>` to create the object since it requires initialisation of
+the root. Such an object is created as follows: `BinaryTree<Gate> obj = new BinaryTree(new OrGate(1,1))`. This sets the root
+of the Binary Tree to the object specified.
+
+#### Using `BinaryTree#isNullAtIndex(int)`
+`BooleanTemplate` uses this function to render the current configuration of the circuit in a String format. The method
+is also extensively used in other internal operations in `BinaryTree<T>` for checking whether a position in the tree has been set or not.
+
+#### Using `BinaryTree#getParentIndex(int, T)`
+Similar to `isNullAtIndex(int)`, this method is used in rendering the current configuration of the circuit in String format.
+
+#### Using `BinaryTree#insert(int, T)`
+In order to enable the ability to populate the `ArrayList<T>` at any node which has a non-null parent node
+the `ArrayList<T> arrayList` attribute is pre-populated with 15 `null` values. The same attribute is modified in the insert()
+operation in the list. Since `insert(int, T)` makes use of `ArrayList<T>.set(int, T)`, values in `arrayList` can be overwritten with this function.
+
+The following sequence diagram is a depiction of the events succeeding a call to `insert(1, new AndGate(1,1))`. 
+![InsertSequenceDiagram](diagrams/BinaryTreeInsertSequenceDiagram.png)
+
+Post calling this function, the second element in the `arrayList` will be the `AndGate(1,1)` object.
+
+#### Using `BinaryTree#isLeaf(int)`
+This function is used by the Logic class `BooleanTemplate` to calculate output values in the digital circuit. It 
+returns whether the node at the input index is a leaf node or not.
+
+The following sequence diagram is a depiction of the events succeeding a call to `isLeaf(1)`on the current `arrayList`:
+
+![IsLeafSequenceDiagram](diagrams/BinaryTreeIsLeafSequence.png)
+
+#### Using `BinaryTree#isEmpty`
+This is used by `BooleanTemplate` to ensure no calculations are being performed on an empty tree.
+
+The following sequence diagram showcases the events succeeding a call to `isEmpty()` on the current object of the `BinaryTree<Gate>` 
+
+![IsEmptySequenceDiagram](diagrams/BinaryTreeIsEmptySequenceDiagram.png)
+ 
+### Rendering Current Boolean Circuit State
+Using a _standard I/O operation_ (Like _Sopln()_) on an object of the `BooleanTemplate` class yields the current configuration
+of the system.
+Each node of the system is represented by a signal ranging from B to O (OUT being the root). All nodes with null parent nodes are
+not shown in the diagram.
+
+## Product scope
+### Target user profile
```

Comment:
Should be
```suggestion
## Appendix: Requirements
### Product scope
#### Target user profile
```

Reply:
Oh no oops, perhaps I'll add it in a next PR?